src/jit/flowgraph.cpp

```diff
@@ -2585,171 +2585,91 @@ void Compiler::fgComputeDoms()
-    // assign numbers to the non-reachable ones, just assign incrementing
-    // values.  We must reach fgBBcount at the end.
+    // If the imaginary root is present then we'll need to create a forest instead of a tree.
+    // Forest roots are chained via DomTreeNode::nextSibling and we keep track of this list's
+    // tail in order to append to it. The head of the list if fgFirstBB, by construction.
+    BasicBlock* rootListTail = fgFirstBB;
 
-    for (i = 1; i <= fgBBNumMax; ++i)
+    // Traverse the entire block list to build the dominator tree. Skip fgFirstBB
+    // as it is always a root of the dominator forest.
+    for (BasicBlock* block = fgFirstBB->bbNext; block != nullptr; block = block->bbNext)
     {
-        if (BlockSetOps::IsMember(this, domTreeEntryNodes, i))
-        {
-            if (domTree[i] == nullptr)
-            {
-                fgDomTreePreOrder[i]  = preNum++;
-                fgDomTreePostOrder[i] = postNum++;
-            }
-        }
-    }
+        BasicBlock* parent = block->bbIDom;
 
-    noway_assert(preNum == fgBBNumMax + 1);
-    noway_assert(postNum == fgBBNumMax + 1);
-    noway_assert(fgDomTreePreOrder[0] == 0);  // Unused first element
-    noway_assert(fgDomTreePostOrder[0] == 0); // Unused first element
-
-#ifdef DEBUG
-    if (0 && verbose)
-    {
-        printf("\nAfter traversing the dominance tree:\n");
-        printf("PreOrder:\n");
-        for (i = 1; i <= fgBBNumMax; ++i)
+        if (parent != imaginaryRoot)
         {
-            printf(FMT_BB " : %02u\n", i, fgDomTreePreOrder[i]);
+            assert(block->bbNum < bbArraySize);
+            assert(parent->bbNum < bbArraySize);
+
+            domTree[block->bbNum].nextSibling = domTree[parent->bbNum].firstChild;
+            domTree[parent->bbNum].firstChild = block;
         }
-        printf("PostOrder:\n");
-        for (i = 1; i <= fgBBNumMax; ++i)
+        else if (imaginaryRoot != nullptr)
         {
-            printf(FMT_BB " : %02u\n", i, fgDomTreePostOrder[i]);
-        }
-    }
-#endif // DEBUG
-}
-
-BlockSet_ValRet_T Compiler::fgDomTreeEntryNodes(BasicBlockList** domTree)
-{
-    // domTreeEntryNodes ::  Set that represents which basic blocks are roots of the dominator forest.
+            assert(rootListTail->bbNum < bbArraySize);
 
-    BlockSet domTreeEntryNodes(BlockSetOps::MakeFull(this));
+            domTree[rootListTail->bbNum].nextSibling = block;
+            rootListTail                             = block;
 
-    // First of all we need to find all the roots of the dominance forest.
-
-    for (unsigned i = 1; i <= fgBBNumMax; ++i)
-    {
-        for (BasicBlockList* current = domTree[i]; current != nullptr; current = current->next)
-        {
-            BlockSetOps::RemoveElemD(this, domTreeEntryNodes, current->block->bbNum);
+            // Clear the imaginary dominator to turn the tree back to a forest.
+            block->bbIDom = nullptr;
         }
```

Comment:
`else unreached()`? Can we have a block that is not `fgFirstBB` with `bbIDom == nullptr`?

Reply:
Yes, SSA dominance produces disconnected nodes because it does not add an imaginary entry like `fgComputeDoms`. Note that the old SSA dominator tree building code did have a `bbIDom` null check but with a somewhat incorrect/misleading comment:
```C++
    // bbIDom for (only) fgFirstBB will be NULL.
    if (bbIDom == nullptr)
    {
        return;
    }
```
The problem is what blocks are traversed to build the tree. In SSA builder there were 2 `ComputeDominators` functions:
1. `void SsaBuilder::ComputeDominators(Compiler* pCompiler, BlkToBlkVectorMap* domTree)` traversed the block list
2. `void SsaBuilder::ComputeDominators(BasicBlock** postOrder, int count, BlkToBlkVectorMap* domTree)` traversed the postOrder array generated by `TopologicalSort` and that only includes unreachable blocks.
The comment was correct in case 2 but not in case 1 as in that case some unconnected blocks were traversed and those have `null` `bbIDom`.

Now `fgBuildDomTree` always does 1 so it has to deal with null `bbIDom` by either ignoring them like SSA did or adding them as forest roots like the old `fgBuildDomTree` did for numbering purposes.

As I mentioned in the PR description - initially I hoped to do a bit more in the dominance area but some things are rather weird or at least confusing. And even tree building, which normally would be pretty trivial, has a few gotchas.